src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNIndexCache.java

```diff
@@ -320,22 +339,52 @@ public KNNIndexCacheEntry loadIndex(String indexPathUrl, String indexName) throw
      * upon expiration of the cache.
      */
     private static class KNNIndexCacheEntry {
+
         private final KNNIndex knnIndex;
+        private final KNNFIndex knnFindex;
         private final String indexPathUrl;
         private final String esIndexName;
         private final WatcherHandle<FileWatcher> fileWatcherHandle;
 
         private KNNIndexCacheEntry(final KNNIndex knnIndex, final String indexPathUrl, final String esIndexName,
                                    final WatcherHandle<FileWatcher> fileWatcherHandle) {
             this.knnIndex = knnIndex;
+            this.knnFindex = null;
```

Comment:
Seems to me that knnIndex and knnFindex should inherit from some kind of abstract class.

Reply:
> I see. I think we should add a different file extension for faiss graphs as opposed to nmslib graphs. Maybe .faiss_hnsw. 

in the latest commit, i use `knnEngine` name in the index file names like:
```
String hnswFileName = String.format("%s_%s_%s%s", 
    state.segmentInfo.name, 
    knnEngine.getLatestBuildVersion(),
    field.name, 
    KNNCodecUtil.HNSW_EXTENSION);
```

> Seems to me that knnIndex and knnFindex should inherit from some kind of abstract class.

yes, you are right. let me think about how to abstract this.

Reply:
> Seems to me that knnIndex and knnFindex should inherit from some kind of abstract class.

I used KNNIndex as abstract class. `KNNFaissIndex` and `KNNNmsLibIndex` extends the KNNINdex. 
`KNNIndexCacheEntry` only save the KNNIndex.

if you have any other idea, pls let me know

Reply:
I think that makes sense. Thanks for making the change!